2010-10-12  Michael Tautschnig  <mt@debian.org>

	* setup-storage/Commands.pm: Use proper dependencies when entire disks are
		used as part of a RAID or LVM volume (thanks Peter Kruse for reporting).
Index: trunk/lib/setup-storage/Commands.pm
===================================================================
--- trunk.orig/lib/setup-storage/Commands.pm
+++ trunk/lib/setup-storage/Commands.pm	
@@ -183,7 +183,9 @@
   # set the raid/lvm unless this is an entire disk flag
   my $cmd = "parted -s $disk set $part_no $t on";
   $cmd = "true" if ($part_no == -1);
-  &FAI::push_command( $cmd, "cleared2_$disk,exist_$d", "type_${t}_$d" );
+  my $pre = "exist_$d";
+  $pre .= ",cleared2_$disk" if (defined($FAI::configs{"PHY_$disk"}));
+  &FAI::push_command( $cmd, "$pre", "type_${t}_$d" );
   if (defined($FAI::partition_table_deps{$disk})) {
     $FAI::partition_table_deps{$disk} .= ",type_${t}_$d";
   } else {
@@ -324,12 +326,15 @@
         }
 
         $d = &FAI::enc_name($d);
-	if ($vol->{preserve}) {
-	  $pre_req .= (&FAI::phys_dev($d))[0] ?
-	      ",pt_complete_" . (&FAI::phys_dev($d))[1] :
-	      ",exist_$d";
+        my ($i_p_d, $disk, $part_no) = &FAI::phys_dev($d);
+        if ($vol->{preserve}) {
+          $pre_req .= ($i_p_d && defined($FAI::configs{"PHY_$disk"})) ?
+            ",pt_complete_$disk" :
+            ",exist_$d";
         } elsif (&FAI::set_partition_type_on_phys_dev($d, "raid")) {
-          $pre_req .= ",pt_complete_" . (&FAI::phys_dev($d))[1];
+          $pre_req .= defined($FAI::configs{"PHY_$disk"}) ?
+            ",pt_complete_$disk" :
+            ",exist_$d";
         } else {
           $pre_req .= ",exist_$d";
         }
